fixes #13618 - cache expensive vmware api calls

timogoebel opened the pull request:
In my dev environment the `compute_resource_selected` call got approx. 10 times faster. I'm really thrilled!

The caching can be enabled for every compute resource. The code is reusable to plugins or other compute resources can reuse the caching mechanism.
Cache invalidation was not straight forward to implement because I wanted to support memcached so I had to use a "scope" when creating the cache keys. The scope is changed when the cache is invalidated.

Some "benchmarks":

Before:
Completed 200 OK in 16121ms (Views: 15963.0ms | ActiveRecord: 6.4ms)

After:
Completed 200 OK in 1710ms (Views: 1583.8ms | ActiveRecord: 5.1ms)

timogoebel committed:
fixes #13618 - cache expensive vmware api calls

tbrisker commented on app/services/compute_resource_cache.rb:
```diff
@@ -0,0 +1,82 @@
+
+  def initialize(compute_resource, opts = {})
+    self.compute_resource = compute_resource
+    self.cache_duration = opts.fetch(:cache_duration, 180.minutes)
```
You can use a kwarg here

ohadlevy replied:
would it make sense to include some sort of a hourly task that keeps refreshing the cache offline?

timogoebel replied:
Yes, I do want to add a hourly rake task for this that can be called by cron. I'd prefer to do this in a separate PR, though.

tbrisker commented on app/controllers/api/v2/compute_resources_controller.rb:
```diff
@@ -162,12 +162,24 @@ def associate
           when 'available_images', 'available_clusters', 'available_flavors', 'available_folders', 'available_networks', 'available_resource_pools', 'available_security_groups', 'available_storage_domains', 'available_zones', 'associate', 'available_storage_pods'
             :view
+          when 'invalidate_cache'
+            :edit
```
I think it makes sense to allow viewers to also invalidate the cache (for example if they notice the data seems outdated), this doesn't actually do anything to the compute resource other then reload the data.

tbrisker commented on app/controllers/compute_resources_controller.rb:
```diff
@@ -123,7 +137,7 @@ def resource_pools
   def action_permission
     case params[:action]
-      when 'associate'
+      when 'associate', 'invalidate_cache'
```
ditto

tbrisker commented on app/models/concerns/compute_resource_caching.rb:
```diff
@@ -0,0 +1,13 @@
+  private
+
+  def cache
+    @cache || ComputeResourceCache.new(self)
```
I think this needs to be `||=` for the new cache to be saved

tbrisker commented on app/views/compute_resources/form/_vmware.html.erb:
```diff
@@ -7,4 +7,6 @@
 <%= checkbox_f f, :set_console_password, :checked => f.object.set_console_password?,
                   :label => _("Console passwords"),
                   :help_inline => _("Set a randomly generated password on the display connection") %>
+<%= checkbox_f f, :caching_enabled, :label => _("Enable caching"),
```
This parameter should also be added to the api

tbrisker commented on db/migrate/20170127094357_add_caching_enabled_to_compute_resource.rb:
```diff
@@ -0,0 +1,5 @@
+class AddCachingEnabledToComputeResource < ActiveRecord::Migration
+  def change
+    add_column :compute_resources, :caching_enabled, :boolean
```
This should have a default false I believe

ohadlevy replied:
do you see a need to default to false? I would rather opt in by default, no one wants the slower option unless they have a problem?

tbrisker reviewed (changes requested):
Looks very good in general, a few comments inline about the code. Couldn't test as I don't have a vmware resource.

ohadlevy commented on app/views/compute_resources/show.html.erb:
```diff
@@ -4,6 +4,7 @@
 <% title_actions display_link_if_authorized(_("Associate VMs"),
                                             hash_for_associate_compute_resource_path(:compute_resource_id => @compute_resource).merge(:auth_object => @compute_resource, :permission => 'edit_compute_resources'),
                                             :title => _("Associate VMs to Foreman hosts"), :method => :put, :class=>"btn btn-default"),
+(display_link_if_authorized(_('Clear Cache'), hash_for_invalidate_cache_compute_resource_path(@compute_resource).merge(:auth_object => @compute_resource), :method => :put, :class => "btn btn-default") if @compute_resource.respond_to?(:invalidate_cache) && @compute_resource.caching_enabled),
```
nitpick: I would call it reload vs clear cache.

timogoebel replied:
How about "refresh cache"?

ohadlevy commented on app/views/compute_resources/show/_vmware.html.erb:
```diff
@@ -6,3 +6,7 @@
 </tr>
+<tr>
+  <td><%= _("Caching") %></td>
+  <td><%= @compute_resource.caching_enabled ? _("Enabled") : _("Disabled") %></td>
```
consider using an icon ? from http://www.patternfly.org/styles/icons/#_ maybe  pficon-ok and  pficon-warning-triangle-o?

timogoebel replied:
I actually do prefer the text vs. the icons to be consistent in the UI. We can change this to icons in a separate PR. `pficon-warning-triangle-o` is too much warning. If a user disables the caching, it won't break anything.

ohadlevy commented on config/routes.rb:
```diff
@@ -399,6 +399,7 @@
           get 'resource_pools'
           post 'ping'
           put 'associate'
+          put 'invalidate_cache'
```
nitpick: i would prefer reload vs invalidate (sounds scary as we all know cache invalidation is a hard problem ;))

ohadlevy commented:
I would like to see a refresh / reload button on the host edit page, i think this is where you would really need  up to date / latest info as you dont have the option (dc, cluster etc).

timogoebel commented:
> I would like to see a refresh / reload button on the host edit page, i think this is where you would really need up to date / latest info as you dont have the option (dc, cluster etc).

@ohadlevy : Would it be ok for you to add this in a separate PR? I think, this will make coding and reviewing easier.

tbrisker commented on app/services/compute_resource_cache.rb:
```diff
@@ -0,0 +1,89 @@
+
+  delegate :logger, :to => ::Rails
+
+  def initialize(compute_resource:, cache_duration: nil)
```
I'm afraid required kwargs is supported only in ruby 2.1, so we can't use that yet. I'd suggest making the first parameter a regular argument and just the cache duration a kwarg. Then you can set the default to 180.minutes in the kwarg rather then below, like so:
```ruby
def initialize(compute_resource, cache_duration: 180.minutes)
```

tbrisker commented on db/migrate/20170127094357_add_caching_enabled_to_compute_resource.rb:
```diff
@@ -0,0 +1,5 @@
+class AddCachingEnabledToComputeResource < ActiveRecord::Migration
+  def change
+    add_column :compute_resources, :caching_enabled, :boolean, default: true
```
Won't this make compute resources that don't support caching appear to have caching enabled?

timogoebel replied:
@tbrisker : Yes, it will. But it's not shown in the UI or API. So it figured, it doesn't really matter.   Do you see that differently?

tbrisker replied:
ok, makes sense.

tbrisker reviewed (commented):
A couple more comments inline. As for the naming of the action, "reload" seems to be a bit misleading, as it isn't actually reloaded, rather "cleared" or "flushed" IIUC.

tbrisker commented on app/services/compute_resource_cache.rb:
```diff
@@ -0,0 +1,89 @@
+
+  def initialize(compute_resource, cache_duration: nil)
+    self.compute_resource = compute_resource
+    self.cache_duration = cache_duration || 180.minutes
```
Instead of defaulting the kwarg to nil then defaulting here to 180.minutes, you can do that in the kwarg itself

timogoebel commented:
@tbrisker , @ohadlevy : Addressed all of your comments. Thanks a lot for these.

> As for the naming of the action, "reload" seems to be a bit misleading, as it isn't actually reloaded, rather "cleared" or "flushed" IIUC.

@tbrisker : Would `refresh cache` be fine for your opinion? I had `clear cache` before, but @ohadlevy suggested to use something different.

I currently see these follow up tasks:

- [ ] Add a rake task to refresh the cache.
- [ ] Add a cache refresh button to the select boxed in the host form that flushes the cache and reloads the data.

tbrisker commented:
@timogoebel I guess refresh is better, @ohadlevy what do you think? Still one minor comment on the kwargs, other than that LGTM but I would like someone with access to VMWare to test it as well before merging.

timogoebel commented:
@shlomizadok : Would you mind testing this for me? I think a VM show and create should suffice.

timogoebel commented:
Caching is now disabled if the compute resource has not been saved yet, fixed the kwarg default and removed caching from `dc` method as method is quite fast.

timogoebel commented:
Rebased and renamed `reload_cache` to `refresh_cache`.

timogoebel commented:
Fixed an issue with caching of `dc.clusters`.

@TheKangaroo: Would you mind testing this PR and let us know if you see any issues? Thanks.

tbrisker commented on app/models/compute_resources/foreman/model/vmware.rb:
```diff
@@ -53,41 +54,54 @@ def max_memory
       if opts[:storage_domain]
-        name_sort(dc.datastores.get(opts[:storage_domain]))
+        cache.cache(:"datastores-#{opts[:storage_domain]}") do
+          name_sort(dc.datastores.get(opts[:storage_domain]))
```
not directly required, but this could benefit from a kwarg for storage domain

tbrisker commented:
Looks good to me pending testing, test failure is unrelated.

TheKangaroo commented:
Hi @tbrisker, I tested it in our environment.
I couldn't find any issues and the runtime improvements in creating hosts is huge :)
I really like your PR @timogoebel

shlomizadok commented:
@tbrisker - testing this now. give me 2 minutes to ack :)

shlomizadok commented:
ACK :+1:

tbrisker commented:
Thanks @timogoebel !

timogoebel commented:
Created two follow-up tickets:

Cache-Warming rake task:
http://projects.theforeman.org/issues/18314

Refresh from host page:
http://projects.theforeman.org/issues/18315
